### ISSUE TITLE
Enable animated rendering of the series

### ISSUE THREAD

**Issue by jedrus07** [ScottLogic/d3fc#601]
It is very easy to get an animated rendering of a time series and it looks really impressive. An example of what I mean can be found in the primefaces showcase:

http://www.primefaces.org/showcase/ui/chart/animate.xhtml

All that is needed on the implementation side is something that masks already rendered graph, giving the effect of animation.

**Comment by chrisprice** [comment]
I think you can get a similar effect already with decorate -

```js
var candlestick = fc.series.candlestick()
    .decorate(function(g) {
        g.transition()
            .duration(1000)
            .ease('linear')
            // hijack the implicit opacity transition
            .tween('opacity', function(d, i) {
                return function(t) {
                    this.style.opacity = (t >= i / g.size()) ? '1' : '0';
                };       
            });
    });
```

https://jsbin.com/tomimutogo/1/edit?html,js,output

**Comment by bjedrzejewski** [comment]
Yes, it works nice for candlestick, but I think this does not work for  <code>.line()</code> and I am not sure if this would work nicely for bar charts. Also try <code>var data = fc.data.random.financial()(10);</code>. I think there is some value in adding a simple <code>.animate(time)</code> which can be very simple to use and is something that people usually like.

**Comment by chrisprice** [comment]
That's a fair point about `line` and the appearance for small numbers of bars. Although I do think it would be possible to get the right effect by using `decorate` to animate `y` for bars. 

My preference would be to use `decorate` to add this functionality if we can. That's not to say that we can't create a component which can be used as a decorate function though e.g. something like (I know this isn't quite right and comes with its own problems but...) -

```js
var wipe = function(g) {
  g.each(function() {
    var defs = d3.select(this.parentNode)
      .selectAll('defs')
      .data([null]);
    defs.enter()
      .append('defs')
      .append('clipPath')
      .attr('id', 'wipe-mask')
      .append('rect')
      .attr({ x: '-100%', width: '0%', height: '100%'});
    defs.select('clipPath')
      .select('rect')
      .transition()
      .ease('linear')
      .duration(1000)
      .attr('width', '200%');
    d3.select(this)
      .attr('clip-path', 'url(#wipe-mask)');
  });
};

// ...
var multi = fc.series.multi()
    .series([gridlines, candlestick])
    .decorate(wipe);
chart.plotArea(multi);
```

https://jsbin.com/nizebevaya/2/edit?html,js,output

By implementing the functionality with `decorate` we're not mandating a particular animation and we expose the `wipe` component for other use-cases.

**Comment by bjedrzejewski** [comment]
I like the decorate idea- I agree that this is way cleaner than the 'hacky solution'. Would you agree that it is still worth packaging some nice 'decorate' usages to get that effect behind the <code>.animate()</code> for the users?

**Comment by chrisprice** [comment]
Yes I think a solid implementation of just the one effect (e.g. wipe) would be a good starting point. After we've had a chance to try that out in a few situations, we could start to consider variations or maybe even a generalised solution. 

I'd be keen to avoid creating a big package of them without having that period of reflection first. We've been in that situation before and it can lead to a lot of code and documentation effort which is ultimately superseded when a more elegant solution comes along.
